connector proxy - materializations

jixij opened the pull request:
**Description:**
The PR adds a Rust crate of `connector-proxy` and changed the materialize connectors to work with it.
Related [issue](https://github.com/estuary/connectors/issues/102)

A connector proxy delegates the communications from Flow Runtime to the connectors, and allows additional functionalities to be executed in the form of plugins during the delegation. Its immediate benefit is to enable the network-proxy for all connectors that are delegated by the network proxy.

The key components of the connector proxy includes the `proxy command`, `plugins`, and `connector runners`, see [this](https://github.com/estuary/flow/pull/379/files#diff-e04a48ac9053a8c454e64e9d750731c6eb2f4c81ba3be41f2a58e69b70f49bb0R96-R104) for detailed explanations.

In this PR, the connector proxy if enabled for materialize connectors. The connector proxy can be enabled without any changes to the existing docker images. To run the images through the connector proxy, the Flow Runtime does the following:
- Pull the connector image to local, find its entrypoint, and other related information via image labels.
- Copy the `flow-connector-proxy` to a temp dir in the host, and mount it to the connector image during execution.
- Start the connector image with entrypoint being `flow-connector-proxy`, set the correct protocols, and pass the `connector entrypoint` and other related args for execution to the `docker run` command.

After the connector proxy is started, it will invoke the original connector binary via the given connector entry point as a separate process, and delegate / modify the communications to it.

Next steps:
- Enable the airbyte protocol in the connector proxy, which involves a) implementing the detailed airbyte protocol inside the [airbyte_source_connector_running](https://github.com/estuary/flow/pull/379/files#diff-5321cf99d20e510250997caba24e0777ada38fd321f5d02e090cc3d6ebd892c4R18-R19), and enable the connector proxy in the go/capture/driver/airbyte/driver.go
- Maybe enable a Label in the connectors images to allow optionally pick the plugins to run for the connector.

**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**
After this change, the network proxy could be enabled for all `materialize connectors`.  Its config is located at the top-level key of `networkProxy`,  [This](https://github.com/estuary/flow/pull/379/files#diff-70a7e23786039866157b5ef60f0fc581baf7c29c0d5fdb440312ef7545cf4f0dR6-R13) is an example of connector config specifications.
All the `sshforwarding` fields are the same as what the `source-postgres` is current using, except that the [spec](https://github.com/estuary/connectors/blob/1c4c927a35e25c85085377310282a2a16c3c0dab/network-proxy-service/networkproxy.go#L24) of `"proxyType": "sshForwarding"` is no longer needed. The presence of `sshforwarding` config is enough to tell the type of network proxies needed.

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)


**Notes for reviewers:**

(anything that might help someone review this PR)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/379)
<!-- Reviewable:end -->

jgraettinger reviewed (changes requested):
Thanks @jixij! Some comments below. Let me know what's unclear, or if you'd like to talk through any of these.

> Reviewed 7 of 7 files at r1, 15 of 15 files at r2, 8 of 8 files at r3, 11 of 11 files at r4, 13 of 13 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/379#-:-MwYEFMP7qVYIVrE1zbU:b-qw0h2f)* status: all files reviewed, 14 unresolved discussions (waiting on @jixij and @psFried)

___
*[go.mod, line 9 at r3](https://reviewable.io/reviews/estuary/flow/379#-MwXpBSGB9XGNKAjfBYx:-MwXpBSGB9XGNKAjfBYy:b-uzyeaz) ([raw file](https://github.com/estuary/flow/blob/4e80d1aca61b03a0beb3727b9b5fb5fb8646f56f/go.mod#L9)):*
> ```XML
> 	github.com/alecthomas/jsonschema v0.0.0-20210818095345-1014919a589c
> 	github.com/bradleyjkemp/cupaloy v2.3.0+incompatible
> 	github.com/containerd/containerd v1.5.9 // indirect
> ```

I'd like to avoid bringing these projects in, and instead continue to shell out as required to `docker` etc.
___
*[crates/connector_proxy/src/errors.rs, line 3 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXXeCb9AYomX5FYb6O:-MwXXeCb9AYomX5FYb6P:b-ifs9p1) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/errors.rs#L3)):*
> ```XML
> #[derive(thiserror::Error, Debug)]
> pub enum Error {
>     #[error("channel send error: {source:?}.")]
> ```

nit: as per the other comment, don't mix `#[from]` and `{source:?}` because it's redundant / already contained within the structured error chain.
___
*[crates/connector_proxy/src/flow_capture_api.rs, line 11 at r5](https://reviewable.io/reviews/estuary/flow/379#-MwXxHLe-KogmArCpi7R:-MwXxHLe-KogmArCpi7S:bpab8tv) ([raw file](https://github.com/estuary/flow/blob/9dcabdfa65d797f41ea8a4d4c6c151203fb1229a/crates/connector_proxy/src/flow_capture_api.rs#L11)):*
> ```XML
> #[allow(unused_variables)]
> #[rustfmt::skip]
> pub trait FlowCapturePlugin: Send + Sync {
> ```

It's not clear to me that this plugin trait modeling is really pulling its weight.

For example, when I look at the wiring related to `do_spec`, it essentially boils down to:

```
        let status =
            invoke_connector(entrypoint, Stdio::inherit(), Stdio::inherit(), "spec")?.wait();
```

Other commands are similar, which means that we have a bunch of fan-out code for each of the various entry points that aren't _really_ differentiated, and which then go on to coalesce back into a common handling path.

An alternative factoring would be bytestream interceptors: functions that are handed metadata about the operation (spec vs validate, etc) and an input stream, and return an output stream. Something like:

```rust
pub type Interceptor = dyn Fn(operation: M, input: Stream<Item=std::io::Result<Bytes>>) -> Box<Stream<Item=std::io::Result<Bytes>>>;
```

See also [tokio-util](https://docs.rs/tokio-util/latest/tokio_util/io/index.html) which has nice conversions between AsyncRead and `Stream<Result<Bytes>>`.

Something like the network proxy can then selectively match against things it knows about and pass through everything else.
You can also trivially chain interceptors by using the output of one as the input of another.
This factoring can also be used for implementing protocol translations, for example from Airbyte source <=> Flow capture.
Note then that you need a _pair_ of Interceptors: one for Flow => Connector, and the other for Connector => Flow.

This read-loop / write-loop framing [is how we implement Airbyte translation](https://github.com/estuary/flow/blob/master/go/capture/driver/airbyte/driver.go#L427-L470) in `connector.Run` today.
___
*[crates/connector_proxy/src/main.rs, line 82 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXca-QATxHvR5wItPx:-MwXca-QATxHvR5wItPy:b-r9vfy7) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/main.rs#L82)):*
> ```XML
> 
> #[derive(Parser, Debug)]
> #[clap(about = "Command to start connector proxies for by flow runtime.")]
> ```

nit: grammar.
___
*[crates/connector_proxy/src/connector_runners/commandutils.rs, line 34 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXkw_TCwbOu69Ch32k:-MwXkw_TCwbOu69Ch32l:bsoq8l3) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/connector_runners/commandutils.rs#L34)):*
> ```XML
> ) -> Result<(), Error> {
>     let mut length_buf: [u8; 4] = [0; 4];
>     if let Err(_) = in_stream.read_exact(&mut length_buf) {
> ```

why is this swallowing / not propagating the error?

Should the cases be:
 * Read zero bytes, then EOF => not an error (e.g, `spec` command).
 * Read a valid message => proxy / not an error.
 * Everything else is an error
___
*[crates/connector_proxy/src/plugins/network_proxy_plugins.rs, line 15 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXof8f639-1q8XlinR:-MwXof8f639-1q8XlinS:bkxoflk) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/plugins/network_proxy_plugins.rs#L15)):*
> ```XML
> 
> impl NetworkProxyPlugin {
>     fn extend_endpoint_schema(endpoint_spec_schema_str: &str) -> Result<String, Error> {
> ```

nit: use serde_json::RawValue here and elsewhere instead of &str, if this is known to be a JSON value?
___
*[crates/connector_proxy/src/plugins/network_proxy_plugins.rs, line 18 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXnw59Dw601YHxV3L5:-MwXnw5A2cAX-RrZy8j0:b-sj1zpz) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/plugins/network_proxy_plugins.rs#L18)):*
> ```XML
>         let network_proxy_schema = create_root_schema::<NetworkProxyConfig>();
> 
>         let mut ond_schema: RootSchema = serde_json::from_str(endpoint_spec_schema_str)?;
> ```

What does "ond" mean?
___
*[crates/network-proxy/Cargo.toml, line 15 at r1](https://reviewable.io/reviews/estuary/flow/379#-MwXVoI9DB3n1HAmkKkJ:-MwXVoI9DB3n1HAmkKkK:bpiysk5) ([raw file](https://github.com/estuary/flow/blob/9bb917b9ee7fe6248c3345b44bca94ee761ad739/crates/network-proxy/Cargo.toml#L15)):*
> ```TOML
> async-trait="*"
> base64="*"
> flow_cli_common = { path = "../flow_cli_common" }
> ```

nit: put this first, separated by whitespace from third-party dependencies.
___
*[crates/network-proxy/src/errors.rs, line 12 at r1](https://reviewable.io/reviews/estuary/flow/379#-MwXV-0C27XYvm20C7g3:-MwXV-0C27XYvm20C7g4:b-ka5a0m) ([raw file](https://github.com/estuary/flow/blob/9bb917b9ee7fe6248c3345b44bca94ee761ad739/crates/network-proxy/src/errors.rs#L12)):*
> ```XML
>     ZeroLocalPort,
> 
>     #[error("SSH error: {source:?}.")]
> ```

When using `#[from]`, don't also put the `source` into the formatted error, as this prints it twice: once in this error's formatting, and again when we traverse the error chain.
___
*[crates/network-proxy/src/sshforwarding.rs, line 27 at r1](https://reviewable.io/reviews/estuary/flow/379#-MwXWObtE3EKzdOtIhjc:-MwXWObtE3EKzdOtIhjd:brh8nl2) ([raw file](https://github.com/estuary/flow/blob/9bb917b9ee7fe6248c3345b44bca94ee761ad739/crates/network-proxy/src/sshforwarding.rs#L27)):*
> ```XML
>     /// User name to connect to the remote SSH server.
>     pub ssh_user: String,
>     /// Base64-encoded private Key to connect to the remote SSH server.
> ```

nit: key shouldn't be capitalized.

This would also be a good place for a short-link that would discuss what this key looks like and how to craft it. I'm unclear on why it's base-64 encoded ... is this a standard `~/.ssh/id_rsa` file?
___
*[crates/network-proxy/src/sshforwarding.rs, line 30 at r5](https://reviewable.io/reviews/estuary/flow/379#-MwY1lJw7CKt_i25IJXk:-MwY1lJw7CKt_i25IJXl:bj7wvwo) ([raw file](https://github.com/estuary/flow/blob/9dcabdfa65d797f41ea8a4d4c6c151203fb1229a/crates/network-proxy/src/sshforwarding.rs#L30)):*
> ```XML
>     pub ssh_private_key_base64: String,
>     /// Host name to connect from the remote SSH server to the remote destination (e.g. DB) via internal network.
>     pub remote_host: String,
> ```

nit: `forward_host` & `forward_port`? "forward" is the terminology used by `man ssh`. Or "tunnel", or "proxy"? "remote" seems ambiguous and conflate-able with `ssh_endpoint` -- at least, I got confused by this for a bit.

Also, it's a bit weird that `ssh_endpoint` is "ssh://host:port", but `remote_host` & `remote_port` are separate configurables. I recognize that `ssh_endpoint` also has the `ssh://` scheme. It's probably also appropriate for the remote address to have a scheme too, since it could be either "tcp://" or a "unix://" socket (even if we don't intend to support "unix://" for now).
___
*[go/connector/proxy.go, line 3 at r3](https://reviewable.io/reviews/estuary/flow/379#-MwXq4Ps-84iuaqDlEMs:-MwXq4Ps-84iuaqDlEMt:b-fbgmwv) ([raw file](https://github.com/estuary/flow/blob/4e80d1aca61b03a0beb3727b9b5fb5fb8646f56f/go/connector/proxy.go#L3)):*
> ```Go
> package connector
> 
> import (
> ```

This is a large departure from `connector.Run` and is unexpected for me. I was assuming a much smaller change to the existing `connector.Run` that threads through CLI arguments for mounting the connector proxy binary into the started docker container. 

I see that you're taking on aspects of inspecting image labels, but I'd rather take a smaller and more incremental step here and would like to avoid bringing in the docker client library or introducing a significant refactoring to how we invoke connectors.

I also would like to keep the Flow runtime as ignorant as possible about the inner details of connectors.
For example, rather than having the Flow runtime determine the entry-point / labels of the image,
I think it should instead pass the output of `docker inspect` into the `connector-proxy` and let it figure it out.
___
*[go/connector/run.go, line 25 at r3](https://reviewable.io/reviews/estuary/flow/379#-MwXr8f3-UdXORwNZbOY:-MwXr8f3-UdXORwNZbOZ:b-2hvktp) ([raw file](https://github.com/estuary/flow/blob/4e80d1aca61b03a0beb3727b9b5fb5fb8646f56f/go/connector/run.go#L25)):*
> ```Go
> )
> 
> type DockerRunCommandBuilder struct {
> ```

As a continuation of my comment from above, I don't think we want a builder pattern here.
This is a significant indirection for readers to follow & understand, which gives up the current concision in `connector.Run` about the concrete CLI arguments which are passed into the invocation.
___
*[tests/sshforwarding/materialize-postgres.ssh.config.yaml, line 13 at r4](https://reviewable.io/reviews/estuary/flow/379#-MwXtAvj6mdMHj2CCpxa:-MwXtAvj6mdMHj2CCpxb:b-e9ah2r) ([raw file](https://github.com/estuary/flow/blob/a7d2f8e5223d72db5be965e36ccfe9ca65d01ebc/tests/sshforwarding/materialize-postgres.ssh.config.yaml#L13)):*
> ```YAML
>     sshEndpoint: ssh://localhost:2222
>     sshUser: flowssh
>     sshPrivateKeyBase64: LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlHNGdJQkFBS0NBWUVBdUF0dmxqWjZqVURldW0zRTRmNVdiMlM5MUlqcVZLN1hUbGN6Y3hVTVBGTjgxSFVNCk9VZmFZOHFyVlpBS2k0Z0p3RGRWZCtmZXRaRXhKdXVlWHZ6TGlBWGJXRWZoZDZ5NnY2T29rdmFMWFlWaDBoM3cKUGxkNnovcUZIcXBJR04za3V0MERWYVhaVVRNRWQ2K1Z4NzcveE53MkN3NEZ1MmJTSS9Lcjl3SXA2aGt1NjlZYgp2RjhlY3lQbGhsS0cvMnZuY1VXMXl0a3ZCT294Umo3QmhNdFlwRkhrZDgwUzBQbHV6a0l4VHphT0JnZHhoYm0wCmdpbDkvQ1FsK0VtWHNBME1vTkVvWWxCcnIyemJzdHc2WXFJWUR2R2k2SnU0Z1BmZ0lRUlpvVm95TzNxYnArZjQKMzhPVEsvRjF4d3FGdzRmWFQ0WlAyS0RQdWJlRG9GNGhBemt6eFg0eXUvNjRZN2RnWkVXbytja3NrU2QydXlFSwplQXZzcnFlTGhtazVZc1ErSFVtbWx0aEV3bHRNMndHemRHRWZNTmxoam1nMjBLNzhJcXczY0ROSmk0cXVkbmsrCkdXUFVic2F4VTRhYmU3OTBiL2VaUktJSzBvZnliUnFGZ3luZ05sbUVHb3AzMGNrQjZwY2g4VHZYaU54MEhIelQKc3hsRzNhTE85TGl6YUtpOUFnTUJBQUVDZ2dHQU4xa1JGTTZMYW9Gd3VnWWE2SS94ZEl5S1hpbFN0MmJoVHlrQQpTd0t3YkUvN1FYVjEyS0FZT0Z5UnpDaUdVMUgxQkw1ZlVMRUZrL3IzR2lwUHRXYzIybEJRbDZyK1Rsdyt5VFV4CmxBamEvcVFIT0FaUzRweXlUV3Vjcncwc28zWFB2TGFNc3l3dVA3SWxETTFjTVBicXVrd3pZSDIyRVphcGFiQW8KTVIrUWRWdGZkZ01zMGlpVTRyR1A3Q1pUZk9yNWJtSTFlcXgzN0hWMW5oZHZZVDA2Q2lpcW1iQ3RrUFVDY3JzTApVelFuMy81TEVKazVoTTl2bU8xU0Vlb24yUVVESFNaNTBUSVFyeFpYUWkrUjRLNlUwUmRudGRwTTJuMjVhS2xUCjRKeW5CaEVrTS9WdHRMbHE0THU1T2Fsa2ZqZjdjVTdKYkVMMHhUVXNpMWMvWXhmL2xya29oV1BoeHIrakdmMmgKbGZiWjUvb1Bid3lJaE1CUHJkSXIvSkJxS0ozUFMwczU1aEVTMDBDNkwzd1RQYnB1clFvUW0zbDB6aTAxaUhQcwpuNzFmS1NTZ1RiU2ordml5d3VnVWUxdWZnU0hoZjBLMVh4djJSWkUrU2R2WGErd21VVmFMRVlyOWhRY0t5cGlhCmpMMWVrVXBXa2kvTmxKZzJmVDNoR040Nm8xR0JBb0hCQU96Lzd3QmVkSVIxRDdwRXp0bWVDcFlJdHpHd3phcVEKd3BHb1RCU29YZ3lscUFTMXZzV3pSdHdnQmlIcys2THFjbmpwdkREbnJFeDY2V1g0Z2t6RHY2VmNSem5TeWhIUQpMa3RITldsbEoxMHpOelhuTkwrSXliZFBwdUozOUdCcXdEMElvWWF2OERrdTBFaEZ1UzJYSmZVNktpU1oyS2FpClJHWmlGLzRVdVloMTRhY0lqbVdBTm9QUXNSYjFVYndtcERKUXpXbDB4L3dvSkkvSEtjekNWK3hyUHdTeDBSbzMKRFk3STl5TXlpU2cxWVQ2dUs3akdBTlFKM1p6TkhBR1VIUUtCd1FER3pLMnRZd0tmUldOMGJrVW1SSDlkTUQ3eAowcFJzcWpmSG5rb3UxNkdzSW0xeWFoSXI4cW1TSG56Q1JJUjdkQ0hkQXlEYm0vcEoxNnNSY2dQWEw2RUJqMGk1CkxMdnRiWXJ5ZU5vb0JMd2FaN3BYeHlKN1kyWWVGMmZCTTVHb0pEWkhnbnRXU1JPdGRUVE9adjZHeHVjbFpyRm0KSGM0TERzbG5xUXVkcnp3Mm45UVJIUDFIS1plTzc1NUg2eVhQb2lHZ3hKSmxMalJQRDYxTXdhZzRnZm1QTnovdApkbmQyZ0VDOUV5TUliNWNveG5UM0tDcVRzN1BueTVoc2tZdVFCU0VDZ2NBU214RVBUVnNHWHc1S0NFYlk3TUJXCkFQOG94b0dPNTJodStQQm9Hc092YUgzbFNNTmVUL0c3TEVSdSsyVFI2R2Z1T0NOUEQ1RUdQQXdBdGw2OFdnMjgKbEo3KzJoNkZBWHZ3NnZrMWVmeDZJNkFzK2lhNDlxc25XZmgzQlFYSlgvZ1JrcERqdXdJMzZ3WHM5cFJRRVBLbwo2dFVPSkluT295U05pUnZyRllOdlJCbmYvQnI1dXZRTXF5Q0w5ZHhoMGdMSDlISDcybkFzeXBYOE41MS9kcTVpCllCVUZjRlJESTVsTDRWa05pVlVkTS91YmFSOERsWC9SSmFraE96VDhwcjBDZ2NCcEtpdVNObmRUcEJSWnFsL0EKUDZvVXRRb0hTUHVsaEdKZm9oZmlNY1hmMXBLSlhSMWh6RXZ4OVZqUU9BalJlZnpUK3V5dDRCU0diYVdjWS9FdQpLNDhEMFRXS1orUytWQjJWNldLV212S2JFVlpkU3dXdHVrWnlFcldiV2RscTFrVTBqUTBMOXpxR255TWhNUkRpCllXUXp2K0NHZ2F2VzQ2cWR3S3BaYnpMdVBweWp0Tzh6YndneTRBV2VUQUhEbGlRdDlPN3d3WUdBZFcyYTFqTlUKT1VXNGdtUzZETnBqZ0xNK2tsb05BMk0yZWMyMFZiWTRUbThVWnZxcUxGQmxEcUVDZ2NBWFFLczV2SFhiWXlFeQpRRnEyY2RSREVyQk9RUWdaVEZUQjBtbFYwTDhBdTk5bFJZV2NQWk5acTdaSW9OZWo4SDh0ZTk1VnVISkN6VVZ2Ckt5U3ZZT2ZpRDh3YVJ1MkdmN0lxQ0hkZ0tCaTdBa0U0NXc3MkdoQytHT29ETk1GZ25sVWdvRGVSenhTN2lkZjQKTUlWUzNzUXplekI3OFpBdVp4MElrSDhQeGdxUkkvRDRDSzlRQkMwYjJJVDF4bXFlNUxDR2hzTUhTdlNjUExWMwpVdTJjczVGa0pVbmtScHd1cDdLRWZKZlpHODBESFA4MUdUc2lvQXQ0MGlneDZnVkFrSW89Ci0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg==
> ```

I think a better user experience would be to _not_ base64 this, eg:

```yaml
networkProxy:
  sshForwarding:
    privateKey: |
      -----BEGIN RSA PRIVATE KEY-----
      Proc-Type: 4,ENCRYPTED
      DEK-Info: AES-128-CBC,47BAB2FE320908911978BB052CD224C1

      blah blah blah
      blah blah blah
      blah blah blah
      blah blah blah
      -----END RSA PRIVATE KEY-----
```

I'm also noting the repetition of "ssh" here. For example, `sshPrivateKey` is redundant given it's already contained within / contextualized by a `sshForwarding` stanza.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go.mod, line 9 at r3](https://reviewable.io/reviews/estuary/flow/379#-MwXpBSGB9XGNKAjfBYx:-Mx4qMef7POzsOaryZsR:b-7kqir0) ([raw file](https://github.com/estuary/flow/blob/4e80d1aca61b03a0beb3727b9b5fb5fb8646f56f/go.mod#L9)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

I'd like to avoid bringing these projects in, and instead continue to shell out as required to `docker` etc.
</blockquote></details>

Done. In the revision, docker cmd are used to pull and inspect the image. The inspect results are stored in a JSON file, which is mounted to the running container. The connector-proxy is able to access all the inspect results and parse out the data it needs.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/network-proxy/src/errors.rs, line 12 at r1](https://reviewable.io/reviews/estuary/flow/379#-MwXV-0C27XYvm20C7g3:-Mx4sFvw-kRO3Ybk9HLl:b7x665y) ([raw file](https://github.com/estuary/flow/blob/9bb917b9ee7fe6248c3345b44bca94ee761ad739/crates/network-proxy/src/errors.rs#L12)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

When using `#[from]`, don't also put the `source` into the formatted error, as this prints it twice: once in this error's formatting, and again when we traverse the error chain.
</blockquote></details>

Good point. 
I've removed the nearly all the sources except for the `NetworkProxyError` for connector-proxy. The reason is that the error could occur during catalog build. It is convenient to surfacing the exact error in the message to identify the exact cause of failure (normally it is address aready in use). 


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/network-proxy/src/sshforwarding.rs, line 30 at r5](https://reviewable.io/reviews/estuary/flow/379#-MwY1lJw7CKt_i25IJXk:-Mx4tNFf4jKssxKM2lGM:b-ukjmwz) ([raw file](https://github.com/estuary/flow/blob/9dcabdfa65d797f41ea8a4d4c6c151203fb1229a/crates/network-proxy/src/sshforwarding.rs#L30)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

nit: `forward_host` & `forward_port`? "forward" is the terminology used by `man ssh`. Or "tunnel", or "proxy"? "remote" seems ambiguous and conflate-able with `ssh_endpoint` -- at least, I got confused by this for a bit.

Also, it's a bit weird that `ssh_endpoint` is "ssh://host:port", but `remote_host` & `remote_port` are separate configurables. I recognize that `ssh_endpoint` also has the `ssh://` scheme. It's probably also appropriate for the remote address to have a scheme too, since it could be either "tcp://" or a "unix://" socket (even if we don't intend to support "unix://" for now).
</blockquote></details>

In response to this and following a few comments related to the ssh here:

I've renamed to `forward_*` to avoid confusions. And renamed `ssh_user` to `user` to avoid redundancy.
A few things related :
- I did not renaming `ssh_endpoint` to `endpoint` as I feel it is more confusing.

-  in the underlying thrussh library these are named `host_to_connect` and `port_to_connect`. Not sure it sounds more clear naming in that way?

- It sounds nice to unify the format on the ssh endpoint and forward host. Question is could we also ignore the schema on both of them? (Since for now the schema of ssh endpoint is always "ssh://", and the schema for the forward host is always "tcp://". ). And somewhat related, do we want to support unix forward_endpoint in the future?

- For the spec of PrivateKeyBase64, I make it based-64 encoded b/c it needs to be stored in JSON fields in some stage of the processing.  Although it could be specified in the YAML file, the original key needs to be tweaked to work properly with JSON, as JSON seems not to be working well with multi-line strings.. 
I feel based64-encoded could be easier to communicate in the document (just instruct users to run `cat ${KEY_FILE} | base64 -w 0 `) than detailed instructions to tweak the key file, so I used the base64 encoded.  This is a temp solution, and I can look more into it and see how to support the non-base64 encoded key specification. (I feel it will be much easier with UI support, but haven't thought about it in details.)

Overall, I feel there are some work to be done in this direction, could they be done in a separate PR?


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/proxy.go, line 3 at r3](https://reviewable.io/reviews/estuary/flow/379#-MwXq4Ps-84iuaqDlEMs:-Mx4ypTp1teO3oSzLoC5:b-kgedmt) ([raw file](https://github.com/estuary/flow/blob/4e80d1aca61b03a0beb3727b9b5fb5fb8646f56f/go/connector/proxy.go#L3)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

This is a large departure from `connector.Run` and is unexpected for me. I was assuming a much smaller change to the existing `connector.Run` that threads through CLI arguments for mounting the connector proxy binary into the started docker container. 

I see that you're taking on aspects of inspecting image labels, but I'd rather take a smaller and more incremental step here and would like to avoid bringing in the docker client library or introducing a significant refactoring to how we invoke connectors.

I also would like to keep the Flow runtime as ignorant as possible about the inner details of connectors.
For example, rather than having the Flow runtime determine the entry-point / labels of the image,
I think it should instead pass the output of `docker inspect` into the `connector-proxy` and let it figure it out.
</blockquote></details>

Per previous discussion, I've refact the code for these.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/connector_proxy/src/flow_capture_api.rs, line 11 at r5](https://reviewable.io/reviews/estuary/flow/379#-MwXxHLe-KogmArCpi7R:-Mx5-oxGBD-zn6JTVNqa:b-wk6x0e) ([raw file](https://github.com/estuary/flow/blob/9dcabdfa65d797f41ea8a4d4c6c151203fb1229a/crates/connector_proxy/src/flow_capture_api.rs#L11)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

It's not clear to me that this plugin trait modeling is really pulling its weight.

For example, when I look at the wiring related to `do_spec`, it essentially boils down to:

```
        let status =
            invoke_connector(entrypoint, Stdio::inherit(), Stdio::inherit(), "spec")?.wait();
```

Other commands are similar, which means that we have a bunch of fan-out code for each of the various entry points that aren't _really_ differentiated, and which then go on to coalesce back into a common handling path.

An alternative factoring would be bytestream interceptors: functions that are handed metadata about the operation (spec vs validate, etc) and an input stream, and return an output stream. Something like:

```rust
pub type Interceptor = dyn Fn(operation: M, input: Stream<Item=std::io::Result<Bytes>>) -> Box<Stream<Item=std::io::Result<Bytes>>>;
```

See also [tokio-util](https://docs.rs/tokio-util/latest/tokio_util/io/index.html) which has nice conversions between AsyncRead and `Stream<Result<Bytes>>`.

Something like the network proxy can then selectively match against things it knows about and pass through everything else.
You can also trivially chain interceptors by using the output of one as the input of another.
This factoring can also be used for implementing protocol translations, for example from Airbyte source <=> Flow capture.
Note then that you need a _pair_ of Interceptors: one for Flow => Connector, and the other for Connector => Flow.

This read-loop / write-loop framing [is how we implement Airbyte translation](https://github.com/estuary/flow/blob/master/go/capture/driver/airbyte/driver.go#L427-L470) in `connector.Run` today.
</blockquote></details>

Yes, I have refact the code based on this. 
The `InterceptorStream` is the basic unit to be processed by interceptors. This is more general way to represent the transformations than the plugin structure.

A few notes there -
1. The converter functions of NetworkProxyCaptureInterceptor is not a simple mapping of each value from the input stream. This is because the `ReaderStream` is not segmenting the stream based on protobuf messages, however, the conversion logic needs to parse the stream by messages. Therefore we need to re-arrange the input values from the input stream by messages, which produced another stream using async_stream(https://docs.rs/async-stream/latest/async_stream/).

2. Now the request/response conversion functions are structured in a simple pair. And all the connectors are invoked via the same invoke_connector API. These might be improved / extended in the future, for usecases like  
(a). maintain some state for the interceptors (e.g. have access to the image inspect json file, or keep a logger).
(b). extend the input args for connector invocation, which I suspect might needed for invoking airbyte connectors https://github.com/estuary/flow/blob/master/go/capture/driver/airbyte/driver.go#L163-L165
These will be more clear when start translating the airbyte connector logic.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/connector_proxy/src/connector_runners/commandutils.rs, line 34 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXkw_TCwbOu69Ch32k:-Mx554cH11OFupl1becy:b-794m2s) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/connector_runners/commandutils.rs#L34)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

why is this swallowing / not propagating the error?

Should the cases be:
 * Read zero bytes, then EOF => not an error (e.g, `spec` command).
 * Read a valid message => proxy / not an error.
 * Everything else is an error
</blockquote></details>

I think I was trying to handle a corner-case that could not happen in these streams. And I misunderstood the `read_exact` API, that even when an "ErrorKind::UnexpectedEof" is produced the data in the resulting length_buf is unspecified. All of them are corrected in the revision.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/connector_proxy/src/plugins/network_proxy_plugins.rs, line 18 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXnw59Dw601YHxV3L5:-Mx55qDw1tS_bYU8QGnL:bedvg7s) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/plugins/network_proxy_plugins.rs#L18)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

What does "ond" mean?
</blockquote></details>

Good catch. This is an error incurred during renaming..


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/connector_proxy/src/plugins/network_proxy_plugins.rs, line 15 at r2](https://reviewable.io/reviews/estuary/flow/379#-MwXof8f639-1q8XlinR:-Mx55xlmCExipJMP7TDL:bb6un1j) ([raw file](https://github.com/estuary/flow/blob/000ee5dc40c5d4b8a2abe608cf5e5396253ef9f2/crates/connector_proxy/src/plugins/network_proxy_plugins.rs#L15)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

nit: use serde_json::RawValue here and elsewhere instead of &str, if this is known to be a JSON value?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/run.go, line 25 at r3](https://reviewable.io/reviews/estuary/flow/379#-MwXr8f3-UdXORwNZbOY:-Mx4zAOG7kL7KWP8Okt8:b-896fix) ([raw file](https://github.com/estuary/flow/blob/4e80d1aca61b03a0beb3727b9b5fb5fb8646f56f/go/connector/run.go#L25)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

As a continuation of my comment from above, I don't think we want a builder pattern here.
This is a significant indirection for readers to follow & understand, which gives up the current concision in `connector.Run` about the concrete CLI arguments which are passed into the invocation.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jixij commented:
Thank you Johnny for the in-depth reviews. I've refact the code accordingly. Explanations are provided. There are more work to be done for the network proxy specifications, but I think the current revision is fine to be reviewed again. PTAL. 





<!-- Sent from Reviewable.io -->

jixij commented:
> A quick note on YAML [multi-line strings](https://yaml-multiline.info/): It works using YAML block style, but you may be running afoul of indenting -- the block _must_ be indented under the enclosing property:
> 
> ```
> $ cat << EOF | yq eval -o=json '.' /dev/stdin
> networkProxy:
>   sshForwarding:
>     privateKey: |
>       -----BEGIN RSA PRIVATE KEY-----
>       Proc-Type: 4,ENCRYPTED
>       DEK-Info: AES-128-CBC,47BAB2FE320908911978BB052CD224C1
> 
>       blah blah blah
>       blah blah blah
>       blah blah blah
>       blah blah blah
>       -----END RSA PRIVATE KEY-----
> EOF
> ```
> 
> outputs:
> 
> ```
> {
>   "networkProxy": {
>     "sshForwarding": {
>       "privateKey": "-----BEGIN RSA PRIVATE KEY-----\nProc-Type: 4,ENCRYPTED\nDEK-Info: AES-128-CBC,47BAB2FE320908911978BB052CD224C1\n\nblah blah blah\nblah blah blah\nblah blah blah\nblah blah blah\n-----END RSA PRIVATE KEY-----\n"
>     }
>   }
> }
> ```

Cool, it works well with both `Block Style Indicator` `Indentation Indicator` on yaml, which I was missing before I think. I've made the change. Thank you!

jgraettinger reviewed (changes requested):
This is _really_ shaping up! A couple of questions and notes. I have one remaining comment of substance regarding `connector.Proxy`, which I think can be simplified significantly by integrating it directly into `connector.Run`, as discussed below. Thanks

> Reviewed 44 of 44 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/379#-:-Mx7rn8b6xW-Gi9DVKA8:bp9gef2)* status: all files reviewed, 6 unresolved discussions (waiting on @jixij and @psFried)

___
*[crates/connector_proxy/src/libs/network_proxy.rs, line 79 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7ccna87oGJeZWPyds:-Mx7ccna87oGJeZWPydt:b-mad31y) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/connector_proxy/src/libs/network_proxy.rs#L79)):*
> ```XML
>         tokio::spawn(Self::start_network_proxy(network_proxy_config, rx));
> 
>         // Block for at most 5 seconds for network proxy to be prepared.
> ```

question: we're always waiting five seconds (and hoping that's enough time)? or can it be less?

This may not be possible with the ssh library we're using, but _ideally_ we'd bind the local port synchronously, here, and asynchronously hand it to the network proxy to do its thing. Then we can immediately allow the connector start and initiate connections. They just won't do anything, of course, until the proxy makes more setup progress and begins to call accept() on the local port. Anyway, this if a future improvement.
___
*[crates/connector_proxy/src/libs/stream.rs, line 9 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7fMgXCsqZFLMHNdKK:-Mx7fMgXCsqZFLMHNdKL:b-8hu9iw) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/connector_proxy/src/libs/stream.rs#L9)):*
> ```XML
>     mut reader: R,
> ) -> impl Stream<Item = std::io::Result<Bytes>> {
>     stream! {
> ```

No action required:

There's a copy of data that's fundamentally shouldn't be required here. I had an idea it could be avoided through `StreamReader.into_inner()`, essentially unwrapping it to take back the inner Stream, which could be returned directly.

However, significant bummer: it has an `into_inner` implementation (great!) but, for inexplicable reasons, it doesn't provided access to the held chunk of buffer which is dropped (boo!). If `into_inner` returned that remainder chunk as well than this would be easy (return a stream of the one remaining chunk chained with the underlying stream). [I filed an issue with tokio](https://github.com/tokio-rs/tokio/issues/4552)
___
*[crates/network-proxy/src/errors.rs, line 12 at r1](https://reviewable.io/reviews/estuary/flow/379#-MwXV-0C27XYvm20C7g3:-Mx7je8P5rJfW9j0tHYg:b-45yb3v) ([raw file](https://github.com/estuary/flow/blob/9bb917b9ee7fe6248c3345b44bca94ee761ad739/crates/network-proxy/src/errors.rs#L12)):*
<details><summary><i>Previously, jixij (Jixiang Jiang) wrote…</i></summary><blockquote>

Good point. 
I've removed the nearly all the sources except for the `NetworkProxyError` for connector-proxy. The reason is that the error could occur during catalog build. It is convenient to surfacing the exact error in the message to identify the exact cause of failure (normally it is address aready in use). 
</blockquote></details>

FYI, there's also `#[error(transparent)]`. You don't _have_ to put error text in if you don't have meaningful additional context to provide.
___
*[crates/network-proxy/src/sshforwarding.rs, line 30 at r5](https://reviewable.io/reviews/estuary/flow/379#-MwY1lJw7CKt_i25IJXk:-Mx7k5HN3oiJcMJBQhal:bhu8s7j) ([raw file](https://github.com/estuary/flow/blob/9dcabdfa65d797f41ea8a4d4c6c151203fb1229a/crates/network-proxy/src/sshforwarding.rs#L30)):*
<details><summary><i>Previously, jixij (Jixiang Jiang) wrote…</i></summary><blockquote>

In response to this and following a few comments related to the ssh here:

I've renamed to `forward_*` to avoid confusions. And renamed `ssh_user` to `user` to avoid redundancy.
A few things related :
- I did not renaming `ssh_endpoint` to `endpoint` as I feel it is more confusing.

-  in the underlying thrussh library these are named `host_to_connect` and `port_to_connect`. Not sure it sounds more clear naming in that way?

- It sounds nice to unify the format on the ssh endpoint and forward host. Question is could we also ignore the schema on both of them? (Since for now the schema of ssh endpoint is always "ssh://", and the schema for the forward host is always "tcp://". ). And somewhat related, do we want to support unix forward_endpoint in the future?

- For the spec of PrivateKeyBase64, I make it based-64 encoded b/c it needs to be stored in JSON fields in some stage of the processing.  Although it could be specified in the YAML file, the original key needs to be tweaked to work properly with JSON, as JSON seems not to be working well with multi-line strings.. 
I feel based64-encoded could be easier to communicate in the document (just instruct users to run `cat ${KEY_FILE} | base64 -w 0 `) than detailed instructions to tweak the key file, so I used the base64 encoded.  This is a temp solution, and I can look more into it and see how to support the non-base64 encoded key specification. (I feel it will be much easier with UI support, but haven't thought about it in details.)

Overall, I feel there are some work to be done in this direction, could they be done in a separate PR?


</blockquote></details>

Looks great now!
___
*[crates/network-proxy/src/sshforwarding.rs, line 73 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7kPN6AgXjxAT7MOyr:-Mx7kPN6AgXjxAT7MOys:b-km3zr8) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/network-proxy/src/sshforwarding.rs#L73)):*
> ```XML
>         let local_listen_addr: SocketAddr =
>             format!("127.0.0.1:{}", self.config.local_port).parse()?;
>         self.local_listener = Some(TcpListener::bind(local_listen_addr).await?);
> ```

Related to the comment above, it's looking like it should be possible to bind the port in the network proxy interceptor, and _then_ spawn off the proxy run-loop as well as simultaneously forwarding along the connector stream?

No further action for this PR, I just want to achieve consensus on what would _ideally_ happen.
___
*[go/connector/proxy.go, line 38 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7nhRqA6NFtEpgS16y:-Mx7nhRqA6NFtEpgS16z:b-r6ke7o) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/go/connector/proxy.go#L38)):*
> ```Go
> }
> 
> func NewProxy(logger ops.Logger) (*Proxy, error) {
> ```

I think this can be simplified still further by incorporating the functionality of this `Proxy` type directly into `connector.Run`, which I'd bet will involve less SLOC overall -- and I don't believe any changes to the current `connector.Run` interface would even be required!

* `connector.Run` already creates a temp directory that's intended for bind-mounting into the container. There are some nuances as to where this temp dir can live that relate to docker-in-docker, so best to keep this consolidated to one place.
* `connector.Run` could _always_ do a docker pull & inspect, and inspection results can be added to `jsonFiles` under a fixed name and leverage that existing mechanism.
* It would always copy the connector binary into that temp dir as well.
* As a short-term hack, it could look at the image name and:
  * If it starts with `ghrc.io/estuary/materialize-` then invoke the connector-proxy binary to kick off the container.
  * Otherwise, by-pass the connector-proxy binary and use the current initialization behavior.

Once Airbyte is switched over, we'd remove that hack such that it _always_ invokes through the connector-proxy binary.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/proxy.go, line 38 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7nhRqA6NFtEpgS16y:-MxA4XSgCyysNiNujs3J:b1kgdih) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/go/connector/proxy.go#L38)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

I think this can be simplified still further by incorporating the functionality of this `Proxy` type directly into `connector.Run`, which I'd bet will involve less SLOC overall -- and I don't believe any changes to the current `connector.Run` interface would even be required!

* `connector.Run` already creates a temp directory that's intended for bind-mounting into the container. There are some nuances as to where this temp dir can live that relate to docker-in-docker, so best to keep this consolidated to one place.
* `connector.Run` could _always_ do a docker pull & inspect, and inspection results can be added to `jsonFiles` under a fixed name and leverage that existing mechanism.
* It would always copy the connector binary into that temp dir as well.
* As a short-term hack, it could look at the image name and:
  * If it starts with `ghrc.io/estuary/materialize-` then invoke the connector-proxy binary to kick off the container.
  * Otherwise, by-pass the connector-proxy binary and use the current initialization behavior.

Once Airbyte is switched over, we'd remove that hack such that it _always_ invokes through the connector-proxy binary.
</blockquote></details>

cool. 
I am working on the refact. One question - even after  switching all connectors to run thrugh connector-proxy, the runConnector function still needs the context information to construct `ProxyCommand` (to indicate proxy protocol of flowSource or flowMaterialize).  So feels like it is better to expand the runConnector API to take one additional parameter of `proxyCommand` or `sourceProtocol`, and avoid the image-name-hack from the beginning?


<!-- Sent from Reviewable.io -->

jixij reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/379)* status: all files reviewed, 6 unresolved discussions (waiting on @jgraettinger, @jixij, and @psFried)

___
*[crates/network-proxy/src/sshforwarding.rs, line 73 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7kPN6AgXjxAT7MOyr:-MxA470_1qqIvWGZ3Hjo:b-799b6m) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/network-proxy/src/sshforwarding.rs#L73)):*
> jsonFiles


<!-- Sent from Reviewable.io -->

jgraettinger reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/379)* status: all files reviewed, 6 unresolved discussions (waiting on @jgraettinger, @jixij, and @psFried)

___
*[go/connector/proxy.go, line 38 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7nhRqA6NFtEpgS16y:-MxA7NR3ASGeFKRgqTBp:bgjkyr5) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/go/connector/proxy.go#L38)):*
<details><summary><i>Previously, jixij (Jixiang Jiang) wrote…</i></summary><blockquote>

cool. 
I am working on the refact. One question - even after  switching all connectors to run thrugh connector-proxy, the runConnector function still needs the context information to construct `ProxyCommand` (to indicate proxy protocol of flowSource or flowMaterialize).  So feels like it is better to expand the runConnector API to take one additional parameter of `proxyCommand` or `sourceProtocol`, and avoid the image-name-hack from the beginning?
</blockquote></details>

Hmm. I think I'll take a position that the `connectory-proxy` shouldn't have to be told how to run the connector -- that the "you're a capture" vs "you're a materialization" argument flag shouldn't be something it requires.

It's wrapping a specific connector; there's only *one* way of driving the wrapped connector that's valid, and the proxy can inspect the connector to determine how it should be driven. This can be a combination of a) using labels, and b) looking at the image name prefix (as a shorter-term kludge to remain compatible with existing connector images we've created).


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/connector_proxy/src/libs/network_proxy.rs, line 79 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7ccna87oGJeZWPyds:-MxAiznV24gOAvosFTwB:b-x1vpgx) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/connector_proxy/src/libs/network_proxy.rs#L79)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

question: we're always waiting five seconds (and hoping that's enough time)? or can it be less?

This may not be possible with the ssh library we're using, but _ideally_ we'd bind the local port synchronously, here, and asynchronously hand it to the network proxy to do its thing. Then we can immediately allow the connector start and initiate connections. They just won't do anything, of course, until the proxy makes more setup progress and begins to call accept() on the local port. Anyway, this if a future improvement.
</blockquote></details>

The `timeout` future will complete immediately if `tx` is closed. So during executions, this takes much less time, never wait for as long as 5 secs.

Initially, the network proxy is structured to support go-rust communication (for source postgresql), where the we shell out the rust server in a separate process. And the timeout was set in go-side as a simple mechanism for error detection in rust.

In the connector-proxy, since all code are in Rust, the error propagation will be much easier. As you indicated, with some refact in the network-proxy, we can improve code here to avoid the channel communications for error detection. I leave a TODO here. I'll work on it after we switch all source connectors to network-proxy, and we do not need to worry about the requirements from go side then.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/connector_proxy/src/libs/stream.rs, line 9 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7fMgXCsqZFLMHNdKK:-MxAmR7F0juBnLCMfYMp:b-90zi4d) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/connector_proxy/src/libs/stream.rs#L9)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

No action required:

There's a copy of data that's fundamentally shouldn't be required here. I had an idea it could be avoided through `StreamReader.into_inner()`, essentially unwrapping it to take back the inner Stream, which could be returned directly.

However, significant bummer: it has an `into_inner` implementation (great!) but, for inexplicable reasons, it doesn't provided access to the held chunk of buffer which is dropped (boo!). If `into_inner` returned that remainder chunk as well than this would be easy (return a stream of the one remaining chunk chained with the underlying stream). [I filed an issue with tokio](https://github.com/tokio-rs/tokio/issues/4552)
</blockquote></details>

This sounds great. 
It will be nice to have the access to the remaining chunks, and we can avoid the redundant reading operation in this function.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/network-proxy/src/errors.rs, line 12 at r1](https://reviewable.io/reviews/estuary/flow/379#-MwXV-0C27XYvm20C7g3:-MxB4ZxZCyRppZU1YiD9:b9w53h8) ([raw file](https://github.com/estuary/flow/blob/9bb917b9ee7fe6248c3345b44bca94ee761ad739/crates/network-proxy/src/errors.rs#L12)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

FYI, there's also `#[error(transparent)]`. You don't _have_ to put error text in if you don't have meaningful additional context to provide.
</blockquote></details>

Thanks!. I've tweaked the sigil in `tracing::error` to output debug information, so that we do not need to force the debug formatter in the error message. And also defined transparent on the errors in connector proxy, whose error messages do no provide additional info.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/network-proxy/src/sshforwarding.rs, line 30 at r5](https://reviewable.io/reviews/estuary/flow/379#-MwY1lJw7CKt_i25IJXk:-MxB5B_g5-f62CeUioNH:bd5l6mr) ([raw file](https://github.com/estuary/flow/blob/9dcabdfa65d797f41ea8a4d4c6c151203fb1229a/crates/network-proxy/src/sshforwarding.rs#L30)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

Looks great now!
</blockquote></details>

thanks!


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[crates/network-proxy/src/sshforwarding.rs, line 73 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7kPN6AgXjxAT7MOyr:-MxB5O5J27bKO4ceiY5-:b-plzz1z) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/crates/network-proxy/src/sshforwarding.rs#L73)):*
<details><summary><i>Previously, jixij (Jixiang Jiang) wrote…</i></summary><blockquote>

> jsonFiles
</blockquote></details>

Yes, agree. This will simplify the current logic a lot in the connector proxy, and I will work on that later.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/proxy.go, line 38 at r6](https://reviewable.io/reviews/estuary/flow/379#-Mx7nhRqA6NFtEpgS16y:-MxB5bgl79n1FWb42T65:bppsrkb) ([raw file](https://github.com/estuary/flow/blob/6c2e3a7c13e1e7009c8d6756c71c0bb63110a1f0/go/connector/proxy.go#L38)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

Hmm. I think I'll take a position that the `connectory-proxy` shouldn't have to be told how to run the connector -- that the "you're a capture" vs "you're a materialization" argument flag shouldn't be something it requires.

It's wrapping a specific connector; there's only *one* way of driving the wrapped connector that's valid, and the proxy can inspect the connector to determine how it should be driven. This can be a combination of a) using labels, and b) looking at the image name prefix (as a shorter-term kludge to remain compatible with existing connector images we've created).
</blockquote></details>

As discussed offline, I added an enum to the run api. And I have left a TODO in the connector proxy - we will check the protocol used by runtime v.s. the protocol required by the connector, and will raise an error if they are in-compatible.


<!-- Sent from Reviewable.io -->

jixij commented:
I've refacted the connector.Run as discussed, and left some TODOs in the code. This is ready for another round of review. PTAL!
Thank you again for so much helpful suggestions and feedbacks!





<!-- Sent from Reviewable.io -->

jgraettinger reviewed (approved):
LGTM!

A couple more nits, and I left a more significant comment about how we go about finding the binary which can happen after this PR lands.

> Reviewed 5 of 8 files at r7, 1 of 1 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/379#-:-MxBHjx91UJh1xNz5wZz:b-p5m4sr)* status: all files reviewed, 6 unresolved discussions (waiting on @jgraettinger, @jixij, and @psFried)

___
*[go/connector/run.go, line 24 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBDr1n1wgBzTpHE1tV:-MxBDr1n1wgBzTpHE1tW:b-gkimlb) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L24)):*
> ```Go
> )
> 
> type ConnectorProtocol int
> ```

nit: ConnectorProtocol lives in a package `connector` already and should just be `Protocol`.  See [package names](https://github.com/golang/go/wiki/CodeReviewComments#package-names) .

Supernit: similarly, we're already contextualized by "Flow", so I'd recommend just `Capture` and `Materialize`, but do not feel strongly at all.
___
*[go/connector/run.go, line 31 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBDTUm2a2VyGKIkEmz:-MxBDTUm2a2VyGKIkEn-:b-fxe0at) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L31)):*
> ```Go
> )
> 
> func (c ConnectorProtocol) GetProxyCommand() string {
> ```

nit: this doesn't need to be Exported. Also, per [effective go](https://go.dev/doc/effective_go#Getters) it's not idiomatic for `Get` to be part of a getter's name.
___
*[go/connector/run.go, line 557 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBFQxq0lMW8Gdr3DQQ:-MxBFQxr1NiNsnWs5Q7V:b3jycif) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L557)):*
> ```Go
> func pullImage(ctx context.Context, image string, logger ops.Logger) error {
> 	var combinedOutput, err = exec.CommandContext(ctx, "docker", "pull", image).CombinedOutput()
> 	logger.Log(logrus.InfoLevel, nil, fmt.Sprintf("output from docker pull: %s", combinedOutput))
> ```

nit: This should be Debug or even Trace.
___
*[go/connector/run.go, line 575 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBFvEH9VUkYn1Yq5kM:-MxBFvEH9VUkYn1Yq5kN:b-e16ytw) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L575)):*
> ```Go
> // If envvar FLOW_RUST_BIN is set, it is "${FLOW_RUST_BIN}/flow-connector-proxy".
> // Otherwise, use the default of "/usr/local/bin/flow-connector-proxy".
> const flowConnectorProxy = "flow-connector-proxy"
> ```

I've been wrangling some similar PATH / finding binary issues lately, and just dealt with this related to `sops`. [See this changeset](https://github.com/estuary/flow/pull/390/files#diff-277e014251a8b745c7a7c8b5b84aaef997a46df12844bc188d81c388dff266c6)

I think you want the exact same behavior here: look for a `flow-connector-proxy` binary that's alongside the current executable. If and **only** if one isn't found, then fall back to searching for it on the $PATH.

You can use [LookPath](https://pkg.go.dev/os/exec#LookPath)) for this, which come to think of it, would have been a better approach vis-a-vi using Stat as I did with the `sops` change -- I'll update that in my PR.

With this approach I don't think we want to or need to assume anything about `/usr/local/bin`, or where rust binaries might be installed, etc.

Tell you what, though: this really belongs as a library function, as I'm now aware of three different places that require this behavior. Lemme add that and either you or I can update the usage here after the fact.


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/run.go, line 24 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBDr1n1wgBzTpHE1tV:-MxBno9UC7rJGiRq-gQo:bldzufk) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L24)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

nit: ConnectorProtocol lives in a package `connector` already and should just be `Protocol`.  See [package names](https://github.com/golang/go/wiki/CodeReviewComments#package-names) .

Supernit: similarly, we're already contextualized by "Flow", so I'd recommend just `Capture` and `Materialize`, but do not feel strongly at all.
</blockquote></details>

Done. Thanks a lot! I'll pay more attention to the go style guides


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/run.go, line 31 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBDTUm2a2VyGKIkEmz:-MxBo2_x3MmI2uXCcs6J:b-896fja) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L31)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

nit: this doesn't need to be Exported. Also, per [effective go](https://go.dev/doc/effective_go#Getters) it's not idiomatic for `Get` to be part of a getter's name.
</blockquote></details>

Done!


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/run.go, line 575 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBFvEH9VUkYn1Yq5kM:-MxBpsbj8CPItk11f7oe:b-eexuch) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L575)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

I've been wrangling some similar PATH / finding binary issues lately, and just dealt with this related to `sops`. [See this changeset](https://github.com/estuary/flow/pull/390/files#diff-277e014251a8b745c7a7c8b5b84aaef997a46df12844bc188d81c388dff266c6)

I think you want the exact same behavior here: look for a `flow-connector-proxy` binary that's alongside the current executable. If and **only** if one isn't found, then fall back to searching for it on the $PATH.

You can use [LookPath](https://pkg.go.dev/os/exec#LookPath)) for this, which come to think of it, would have been a better approach vis-a-vi using Stat as I did with the `sops` change -- I'll update that in my PR.

With this approach I don't think we want to or need to assume anything about `/usr/local/bin`, or where rust binaries might be installed, etc.

Tell you what, though: this really belongs as a library function, as I'm now aware of three different places that require this behavior. Lemme add that and either you or I can update the usage here after the fact.
</blockquote></details>

The logic with LookPath in the changeset is very cool. 
I've thought about and was unsure on how to find the proxy binaries, and provided the envvar as a temp solution.
I can update the logic here in the followup PR for the source-connectors once the LoopPath logic gets merged.


<!-- Sent from Reviewable.io -->

jgraettinger reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/estuary/flow/379)* status: 47 of 55 files reviewed, 2 unresolved discussions (waiting on @jgraettinger, @jixij, and @psFried)

___
*[go/connector/run.go, line 575 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBFvEH9VUkYn1Yq5kM:-MxBxOlh-MuU0Vhqgfje:brrn5b6) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L575)):*
<details><summary><i>Previously, jixij (Jixiang Jiang) wrote…</i></summary><blockquote>

The logic with LookPath in the changeset is very cool. 
I've thought about and was unsure on how to find the proxy binaries, and provided the envvar as a temp solution.
I can update the logic here in the followup PR for the source-connectors once the LoopPath logic gets merged.
</blockquote></details>

Just FYI, [it's merged](https://github.com/estuary/flow/blob/master/go/pkgbin/exec.go#L16-L20)


<!-- Sent from Reviewable.io -->

jixij commented:
___
*[go/connector/run.go, line 575 at r11](https://reviewable.io/reviews/estuary/flow/379#-MxBFvEH9VUkYn1Yq5kM:-MxC4IXvCpo2To9msvvc:b-8fx2u7) ([raw file](https://github.com/estuary/flow/blob/e8c0bed3833f61e6d75a0844db43f197a56be9af/go/connector/run.go#L575)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

Just FYI, [it's merged](https://github.com/estuary/flow/blob/master/go/pkgbin/exec.go#L16-L20)
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->
